Generates new token from old token.

amol opened the pull request:
The API Specification http://api.openstack.org/api-ref-identity.html allows users to authenticate and generate token from old token.
Request body format:
    {
        "auth" :
            { "token" : {"id" : "old_token"} }
    }

such request:
    1. validates the token
    2. decodes the token
    3. and generates new

amol committed:
Generates new token from old token.

The API Specification http://api.openstack.org/api-ref-identity.html allows users to authenticate and generate token from old token.
Request body format:
    {
        "auth" :
            { "token" : {"id" : "old_token"} }
    }

such request:
    1. validates the token
    2. decodes the token
    3. and generates new

bodenr commented on jumpgate/identity/drivers/sl/tokens.py:
```diff
@@ -70,11 +70,23 @@ class SLAuthDriver(identity.AuthDriver):
+            token = identity.token_id_driver().token_from_id(token_id)
+            try:
+                identity.token_driver().validate_token(token)
+                username = token['username']
```
the token format is TokenDriver specific and should be only accessed via the token driver. Thus here you should use something like:

``` python
username = identity.token_driver().username(token)
```

bodenr commented on jumpgate/identity/drivers/sl/tokens.py:
```diff
@@ -70,11 +70,23 @@ class SLAuthDriver(identity.AuthDriver):
+            try:
+                identity.token_driver().validate_token(token)
+                username = token['username']
+                credential = token['api_key']
```
this is again token driver specific and thus should not be accessed directly from the token structure.
I think we need to add a method to the `TokenDriver` interface, something like 

``` python
def credential(self, token)
```

which is intended to extract and return the credential from the said token.

we can then impl a concrete version of it in `JumpgateTokenDriver` something like:

``` python
def credential(self, token):
    return token.get('api_key')
```

amol replied:
+1 will add that.

bodenr commented on jumpgate/identity/drivers/sl/tokens.py:
```diff
@@ -70,11 +70,23 @@ class SLAuthDriver(identity.AuthDriver):
+                username = token['username']
+                credential = token['api_key']
+            except exceptions.InvalidTokenError as e:
+                return None
```
maybe best to not 'swallow' this exception here and return none... i would consider just letting the InvalidTokenError propigating up the stack. there is a mapper in `jumpgate.common.exceptions` which will transform this exception into an Unauthorized, but more importanatly it will dump the error to log before doing so to ease admin operations of this service.

bodenr commented on jumpgate/identity/drivers/sl/tokens.py:
```diff
@@ -70,11 +70,23 @@ class SLAuthDriver(identity.AuthDriver):
         credential = lookup(creds, 'auth', 'passwordCredentials',
                             'password')
+        token_id = lookup(creds, 'auth', 'token', 'id')
+        if self._is_token_auth(username, credential, token_id):
```
dont forget that since this is a new form of authN, this if block should return a well formed auth response like that other auth blocks here... e.g.

``` python
            return {'user': user, 'credential': credential,
                    'auth_type': 'api_key'}
```

amol replied:
Need some clarification here. I was assuming, that the token will be used to extract username/password and then rest of the code will be used to return the auth block. 

Are you suggesting something else here?

bodenr replied:
ah ok... sorry my bad -- I misread the implementation so disregard this comment.

sudorandom replied:
Is `_is_token_auth()` required here? Would it not be sufficient to simply check to see if token_id is true-ish with `if token_id:`?

amol replied:
As per the API specification, when user provides token, she should not be giving username/password. If all has been given result should be 400 Bad Request.

amol committed:
Username/Credentials retrived from token driver

Since each token driver is free to use the way it wants to store the
username/password, username() and credential() method are invoked to get
the values. token_driver.credentials() added in view of this need.

bodenr commented:
thanks for handling all the comments... LGTM, +1

amol committed:
Removed _is_token_auth()

If request contains token id it is assumed to be token auth.

amol commented:
@sudorandom and @beittenc Request you to a merge, as all review comments considered.

sudorandom commented:
Thanks @amol
